ambry-router/src/main/java/com.github.ambry.router/CryptoJob.java

```diff
@@ -13,16 +13,36 @@
  */
 package com.github.ambry.router;
 
-import java.security.GeneralSecurityException;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 
 /**
  * CryptoJob representing the job that needs processing by {@link CryptoJobHandler}
  */
-public interface CryptoJob extends Runnable {
+public abstract class CryptoJob<T> implements Runnable {
```

Comment:
why is this an abstract class? I find interfaces more flexible. Abstract classes are less obvious (somebody who overrides needs to remember to call super to get the functionality of the class). A wrapper for callbacks can solve the problem which forced you to make this an abstract class.